source/MRMesh/MRIntersection.h

```diff
@@ -1,21 +1,141 @@
 #pragma once
 
+#include "MRMeshFwd.h"
 #include "MRPlane3.h"
 #include "MRLine3.h"
+#include "MRLineSegm.h"
 #include <optional>
 
 namespace MR
 {
 
+template<typename T>
+const T errorLimit = std::numeric_limits<T>::epsilon() * 20.f;
+
+// finds an intersection between a plane1 and a plane2
+// returns nullopt if they are parallel
+// returns line from plane1 if plane1 equal plane2
+template<typename T>
+std::optional<Line3<T>> intersection( const Plane3<T>& plane1, const Plane3<T>& plane2 )
+{
+    const auto plane1N = plane1.normalized();
+    const auto plane2N = plane2.normalized();
+    const auto crossDir = cross( plane1N.n, plane2N.n );
+
+    const auto planeNpoint = plane1N.n * plane1N.d;
+    const auto plane2Npoint = plane2N.n * plane2N.d;
+
+    if ( crossDir.length() < errorLimit<T> )
+    {
+        if ( ( planeNpoint - plane2Npoint ).length() < errorLimit<T> )
+        {
+            auto bv = plane1N.n.furthestBasisVector();
+            return Line3<T>( planeNpoint, cross( plane1N.n, bv ).normalized() );
+        }
+        else
+            return {};
+    }
+
+    const auto p1 = plane1N.project( plane2Npoint );
```

Comment:
better form the system in Matrix3:
```
p*n_1=d_1
p*n_2=d_2
p*(n_1 x n_2) =0
```
add solve it using Matrix3 methods.
The code will be much cleaner.

Reply:
reworked